### PR TITLE
C#: Clear existing data directory extracted from PCK

### LINKED ISSUE #96299: [Dotnet]Exporting a project in c# after switching aot can be problematic
### Tested versions

v4.4.dev1.mono.official [28a72fa43]

### System information

Godot v4.4.dev1.mono - Arch Linux #1 ZEN SMP PREEMPT_DYNAMIC Tue, 20 Aug 2024 15:33:58 +0000 - Wayland - GLES3 (Compatibility) - Mesa Intel(R) UHD Graphics 630 (CFL GT2) - Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz (8 Threads)

### Issue description

If aot is not enabled for the first export of c# project, and is enabled for the second export, their c# data will be mixed with each other. under linux this will result in the runtime of the second exported game actually using the c# assembly from the first export, not the aot binary from the second export. (windows seems to have a worse problem with this)

### Steps to reproduce

1. Export the project and run it, the screen should display `Hello CSharp!`.
2. Modify `PublishAot` in `TestAot.csproj` to `true`.
3. Use the same export path as the first time, re-export and run. Ideally it should display `Hello CSharp Aot!`. Under Linux, it will actually display `Hello CSharp!`. I have not tested it under Windows.
(When the export option of `embed_build_outputs` is enabled, this problem will still occur even if the two export paths are inconsistent)

### Minimal reproduction project (MRP)

[test-aot.zip](https://github.com/user-attachments/files/16805517/test-aot.zip)

### PR DESCRIPTION
**Pull request by scgm0**
Partial fix for #96299 (embed_build_outputs=true)

### CODE REVIEW

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs [hunk 1]
```diff
@@ -227,6 +227,13 @@ private void _ExportBeginImpl(string[] features, bool isDebug, string path, long
                         projectDataDirName = Path.Combine("Contents", "Resources", projectDataDirName);
                     }
 
+                    // Delete old publish output files to prevent problems with file mixing.
+                    string projectDataDir = Path.Join(ProjectSettings.GlobalizePath("res://"), Path.GetDirectoryName(path), projectDataDirName);
```
Comment (raulsntos) [inline modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:231]: Not sure I trust this to always result in the correct path.

**Review by raulsntos — COMMENTED** [review]
Thanks for the contribution. At a first glance this seems reasonable, but I'm a bit hesitant to deleting directories without some kind of confirmation. It may not be obvious to the user that this data directory is specific to each export and may be overridden, thus breaking the previous export.

We've had similar issues reported in the past about exporting to different platforms/architectures, which is why now the data directory name contains the name of the target platform and architecture. And there's even some users that are not aware that this directory is important and forget to include it when distributing the exported game.

So I'm thinking this may be a documentation issue, and maybe we should recommend that C# projects are always exported to an empty directory.

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs (hunk 1 above)
Comment (scgm0) [reply modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:231]: In my tests, the 'path' is the relative path to the project's 'res' directory as the root (export_presets.cfg stores the 'export_path' is like that), and I just briefly tested the directory exported to 'res' under Linux and a directory that is not 'res', and maybe more testing is needed.

**Comment by paulloz** [comment]
Yeah, I'm not a big fan of nuking folders without user validation either. The export process should 100% be more documented and explicit about all of this. Also, isn't there already a dialog that opens when you export in a non-empty directory? We might be able to add details in there (e.g. saying that "if you're exporting on top of another architecture / aot mode / platform, you might experience unexpected issues" or something).

**Comment by scgm0** [comment]
> Yeah, I'm not a big fan of nuking folders without user validation either. The export process should 100% be more documented and explicit about all of this. Also, isn't there already a dialog that opens when you export in a non-empty directory? We might be able to add details in there (e.g. saying that "if you're exporting on top of another architecture / aot mode / platform, you might experience unexpected issues" or something).

I don't think there's a problem with deleting the old folder, because the folder isn't designed to be the location the user should use, its contents are fixed and created by Godot at export time, and when 'embed_build_outputs' is enabled, it will be directly put into the 'pck' and will not be copied to the local cache directory until the user runs, and whether 'embed_build_outputs' is turned off or on'embed_build_ outputs', it will directly overwrite the previous results,  and there is no convenient way for users to use this folder by themselves, it is created and used by godot automatically.

**Comment by scgm0** [comment]
> Thanks for the contribution. At a first glance this seems reasonable, but I'm a bit hesitant to deleting directories without some kind of confirmation. It may not be obvious to the user that this data directory is specific to each export and may be overridden, thus breaking the previous export.
> 
> We've had similar issues reported in the past about exporting to different platforms/architectures, which is why now the data directory name contains the name of the target platform and architecture. And there's even some users that are not aware that this directory is important and forget to include it when distributing the exported game.
> 
> So I'm thinking this may be a documentation issue, and maybe we should recommend that C# projects are always exported to an empty directory.

If many users are unaware of the importance of this folder, then perhaps the `embed_build_outputs` option should be enabled by default? `embed_build_outputs` is now not only turned off by default, but also hidden in the advanced options. I feel that this will make many users not know that godot has this function.

**Comment by raulsntos** [comment]
> `embed_build_outputs` is now not only turned off by default, but also hidden in the advanced options.

That's because it's a problematic option that causes issues in some platforms and it's likely to be removed in the future.

**Comment by scgm0** [comment]
> > `embed_build_outputs` is now not only turned off by default, but also hidden in the advanced options.
> 
> That's because it's a problematic option that causes issues in some platforms and it's likely to be removed in the future.

I think it works great and a single executable file is more convenient for distribution. What's the problem with it? #95312?

**Comment by raulsntos** [comment]
It's a potential security vulnerability, it leaves garbage in temporary directories that may never be cleaned after removing the game, in some platforms it prevents the exporter from signing the libraries and might attract unnecessary antivirus software attention.

The benefits of having a single-file executable don't outweigh the drawbacks. In some platforms we already have a single-file without this hack because of the nature of how their exporter works (e.g.: macOS exports a bundle, Android exports an APK/AAB). In my opinion, it's not worth having this "feature" which is only relevant to Windows and Linux, and has so many risks.

**Comment by scgm0** [comment]
> It's a potential security vulnerability, it leaves garbage in temporary directories that may never be cleaned after removing the game, in some platforms it prevents the exporter from signing the libraries and might attract unnecessary antivirus software attention.
> 
> The benefits of having a single-file executable don't outweigh the drawbacks. In some platforms we already have a single-file without this hack because of the nature of how their exporter works (e.g.: macOS exports a bundle, Android exports an APK/AAB). In my opinion, it's not worth having this "feature" which is only relevant to Windows and Linux, and has so many risks.

Uh, Android now forces `embed_build_outputs` to be turned on, and it also has the problem in #96299, which is slightly different when `embed_build_outputs` is turned on and when it's not, one happens when the game is launched, and the other happens when the game is export, this pr contains both cases.
Problems during export can be solved by changing the export directory, but problems during startup can only be solved by deleting the old folder.

**Comment by raulsntos** [comment]
I think as long as we still support `embed_build_outputs`, the changes to `godotsharp_dirs.cpp` are fine and will be needed for https://github.com/godotengine/godot/pull/88803, because the changes in that PR make Android exports installed before the PR fail due to an outdated cache (same as https://github.com/godotengine/godot/issues/96299).

But we are still not convinced we want the changes to `ExportPlugin.cs` and that's blocking this PR. Could you separate the changes to `godotsharp_dirs.cpp` into a different PR?

**Comment by scgm0** [comment]
> I think as long as we still support `embed_build_outputs`, the changes to `godotsharp_dirs.cpp` are fine and will be needed for #88803, because the changes in that PR make Android exports installed before the PR fail due to an outdated cache (same as #96299).
> 
> But we are still not convinced we want the changes to `ExportPlugin.cs` and that's blocking this PR. Could you separate the changes to `godotsharp_dirs.cpp` into a different PR?

I can split them into two PRs, but they are both assembly mixing problems, they just happen at different times, and I hope they can be fixed together.

**Comment by scgm0** [comment]
> I think as long as we still support `embed_build_outputs`, the changes to `godotsharp_dirs.cpp` are fine and will be needed for #88803, because the changes in that PR make Android exports installed before the PR fail due to an outdated cache (same as #96299).
> 
> But we are still not convinced we want the changes to `ExportPlugin.cs` and that's blocking this PR. Could you separate the changes to `godotsharp_dirs.cpp` into a different PR?

Uh-huh, if you separate a PR, can you give this PR and the new PR a name?

**Comment by raulsntos** [comment]
> I can split them into two PRs, but they are both assembly mixing problems, they just happen at different times, and I hope they can be fixed together.

I agree that they are very similar, but waiting for both to be fixed at the same time will only delay this PR since we are still undecided about parts of it. I think it'd be better to merge the parts that we can agree on.

> Uh, if split a PR separately, can you come up with a name for this PR and the new PR?

Naming is hard, but I can try.

- For the changes to `ExportPlugin.cs`: `C#: Clear existing data directory in export output`
- For the changes to `godotsharp_dirs.cpp`: `C#: Clear existing data directory extracted from PCK`

**Comment by scgm0** [comment]
> > I can split them into two PRs, but they are both assembly mixing problems, they just happen at different times, and I hope they can be fixed together.
> 
> I agree that they are very similar, but waiting for both to be fixed at the same time will only delay this PR since we are still undecided about parts of it. I think it'd be better to merge the parts that we can agree on.
> 
> > Uh, if split a PR separately, can you come up with a name for this PR and the new PR?
> 
> Naming is hard, but I can try.
> 
>     * For the changes to `ExportPlugin.cs`: `C#: Clear existing data directory in export output`
> 
>     * For the changes to `godotsharp_dirs.cpp`: `C#: Clear existing data directory extracted from PCK`

Thank you!

**Review by raulsntos — APPROVED** [review]
Looks good, thanks!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**